common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -447,6 +447,37 @@ private UTF8String toUpperCaseSlow() {
     return fromString(toString().toUpperCase());
   }
 
+  /**
+   * Optimized lowercase comparison for UTF8_BINARY_LCASE collation
+   */
+  public int compareLowercase(UTF8String other) {
+    int curr;
+    for (curr = 0; curr < numBytes && curr < other.numBytes; curr++) {
+      byte left = getByte(curr);
+      byte right = other.getByte(curr);
+      if (numBytesForFirstByte(left) != 1 || numBytesForFirstByte(right) != 1) {
```

Comment:
This is expensive -- you don't want to know the number of bytes, you just want to know if it's more than 1. If you look at the UTF-8 spec, you see that the multibyte characters all have the high bit set, and the single-byte characters all have the high bit unset. So you could just test for the high bit. Assuming that `toLowerCase` will not go from ASCII to non-ASCII, this also gets rid of the next check in line 463, which is in essence also a test for the high bit.

Reply:
<img width="694" alt="image" src="https://github.com/apache/spark/assets/154237371/a496bd0c-51b7-4fab-b23e-1e80f8108dfd">

you can just check if the byte is positive as java does when creating compact strings

Reply:
Good point @bart-samwel and @stefankandic, adding further optimizations to comparison and case conversions as well since they also contain similar constructs.